Fix for Pumps Turning Off and an Overhaul of the C++ Code in Pumps

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #9427 
 - It was noticed in a particular file that variable speed pumps were totally turning off and this led to no heating for a case when the pump maximum and minimum flow rate were specified by the user.  In the same file, a switch to autosizing those two parameters showed the pumps operating as expected.  The problem was that in the hard-wired input file both the maximum and minimum were exactly the same.  This case was not trapped and resulted in no pump operation.  However, if a user had chosen any other minimum flow rate the file would run with pumps operating and heating reported.  This has now been fixed.  In addition, it was requested that the pump module be significantly cleaned up and this was achieved.  The pump module is much cleaner now.  No real differences in the output files other than a minor difference in the error file due to modifications made to simplify the error messages that are produced.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Correction of Pump Issue Causing No Flow

There was an issue with the logic in the pump get input routine where when the maximum and minimum flow rate were equal the program accepted the input with no warnings but then prevented the pumps from ever running.  This was corrected in the code, an existing unit test was modified, a new unit test was added, and the documentation for the minimum flow rate was clarified.

RKStrand committed:
Pump Updates

Addresses review comments from Amir.  Improved constant declaration and simplification of the entire module using the auto &this... declarations.

RKStrand committed:
Revert "Pump Updates"

This reverts commit 1da1b818837441899fb4cc8b897454c1e9935113.

RKStrand committed:
GetPumpInput VarSpeed only

Changes to the GetPumpInput section that reads variable speed pumps only.  First of many chunks...

RKStrand committed:
GetPumpInput Step 2

Constant speed and condensate pumps

RKStrand committed:
GetPumpInput Step 3

The rest of the "get" part of the routine (not including the output variable calls and beyond).

RKStrand committed:
GetPumpInput last section

Modification of remaining portions of GetPumpInput

RKStrand committed:
InitPump and SetupPumpMinMax

changes to InitPump and SetupPumpMinMaxFlows

RKStrand committed:
Updates in CalcPumps

Continuing the work in CalcPumps

RKStrand committed:
Updates to Remaining Portions

General auto declaration updates to the remaining portions of this file.

RKStrand committed:
Another Iteration on Pumps

More work and a few error fixes for unit tests that were failing.  This moves another step forward and corrects the problems introduced by the last commit.

RKStrand committed:
Pumps Work Formating Error Messages

Reworking of pump error messages to use format.  Also some other updates as requested by @amirroth

RKStrand committed:
Merge branch 'develop' into 9427NoHeatingWithPumpMaxAndMinFlowEqual

RKStrand committed:
Fix Pump Custom Check Errors

A couple of misses with the format statements

RKStrand committed:
Pump Fixed for Remaining Review Comments to Date

This commit includes what hopefully addresses all of the comments received to date from @amirroth.

RKStrand committed:
Final Set of Pre-PR Comments

Resolution of the remaining comments from @amirroth.  Hopefully this comes back clean and is ready to be made into a PR.

RKStrand committed:
MIssed an initialization

Forgot to define this initialization.  Did this cause all of the problems in the ci run?

RKStrand committed:
Revert "MIssed an initialization"

This reverts commit 87366ae5c678105b0840eca78acd4f8e5058c9e8.

RKStrand committed:
Revert "Final Set of Pre-PR Comments"

This reverts commit 03e221e1e672399f0aac05246c1bd8dc51518dd0.

RKStrand committed:
Revert "Revert "Final Set of Pre-PR Comments""

This reverts commit 531418a7d1235e436985c315cda76edbfac9025a.

RKStrand committed:
Revert "Revert "MIssed an initialization""

This reverts commit 7f623b44108d76dd73209afce039a42bd999399a.

RKStrand committed:
Fixes Problems Hopefully

Next time I should pay more attention to my email... Thanks, @amirroth for noticing this and commenting.

amirroth reviewed (approved):
I won't comment on the underlying issue that precipitated this branch and I've made plenty of comments during the restructuring already. It's always possible to do more (spoiler: I am seeing some opportunities now as I write this) but as some point you just have to stop. The usage of old idioms has been reduced as much as it has in any other module that has undergone this sort of pass. I am satisfied. Thanks @RKStrand!

jmarrec commented on tst/EnergyPlus/unit/Pumps.unit.cc:
```diff
@@ -650,9 +650,9 @@ TEST_F(EnergyPlusFixture, VariableSpeedPump_MinFlowGreaterThanMax)
-        "   ** Warning ** GetPumpInput: Pump:VariableSpeed=\"SUPPLY INLET PUMP\", Invalid 'Design Minimum Flow Rate'",
-        "   **   ~~~   ** Entered Value=[2.00000E-003] is above the Design Maximum Flow Rate=[1.00000E-003].",
-        "   **   ~~~   ** Reseting value of 'Design Minimum Flow Rate' to the value of 'Design Maximum Flow Rate'.",
+        "   ** Warning ** GetPumpInput:  Pump:VariableSpeed = SUPPLY INLET PUMP, Invalid 'Design Minimum Flow Rate'",
```
Why this change of punctuation? double space at last (the quotes were also clearer but that's just my two cents)

RKStrand replied:
I personally didn't feel like the quotes added much.  The double space between the GetPumpInput: and Pump:VariableSpeed...that can be gotten rid of.  But I'm also just one opinion.  Doesn't matter to me.  Whatever the team wants I can implement.

jmarrec commented on src/EnergyPlus/Pumps.hh:
```diff
@@ -122,131 +113,93 @@ namespace Pumps {
-              VFDControlType(ControlTypeVFD::Invalid), MaxRPM(0.0), MinRPM(0.0), PumpActualRPM(0.0)
-        {
-        }
+        int ManualRPMSchedIndex = 0;
+        int LowerPsetSchedIndex = 0;
+        int UpperPsetSchedIndex = 0;
+        int MinRPMSchedIndex = 0;
+        int MaxRPMSchedIndex = 0;
+        ControlTypeVFD VFDControlType = ControlTypeVFD::Invalid; // VFDControlType
+        Real64 MaxRPM = 0.0;                                     // Maximum RPM range value - schedule limit
+        Real64 MinRPM = 0.0;                                     // Minimum RPM range value - schedule limit
+        Real64 PumpActualRPM = 0.0;                              // RPM recalculated from final flow through the loop
```
I definitely like the changes in the header. Member init at declaration is much clearer to me.

I wonder how often the Schedule Names are retrieved and used and whether caching (the `std::string xxxSchedName`) was a good idea to begin with. I do not know if this has been discussed / ruled upon before either though.

RKStrand replied:
@amirroth requested this and I ran with it.  The only benefit I see with keeping the schedule name is for debugging, but that's of limited value I think.

amirroth replied:
No, caching schedule names was not a good idea. It has not been discussed in my recollection. PS There are multiple other instances of this type of thing. Too much string copying in general in EnergyPlus.

jmarrec replied:
Thanks for the replies! (I view this as an opportunity to learn)

I agree there's too much caching generally.

For debugging you should be able to do state->Xxxx anyways so that's not a concern.

jmarrec commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -105,19 +105,8 @@ using DataHVACGlobals::ForceOff;
-    {PumpType::Bank_VarSpeed, cPumpBank_VarSpeed},
-    {PumpType::Bank_ConSpeed, cPumpBank_ConSpeed},
-};
+constexpr std::array<std::string_view, static_cast<int>(PumpType::Num)> pumpTypeIDFNames = {
+    "Pump:VariableSpeed", "Pump:ConstantSpeed", "Pump:VariableSpeed:Condensate", "HeaderedPumps:VariableSpeed", "HeaderedPumps:ConstantSpeed"};
```
Won't make a difference on gcc, but I think this should be `static constexpr`

RKStrand replied:
I went with @amirroth's suggestion but if this works and is better, that's fine with me.

amirroth replied:
Hmmm. Good to know. I only use gcc/g++.

jmarrec replied:
@lefticus made a c++ weekly on it the topic recently.

https://youtu.be/4pKtPWcl1Go

(I'm not saying he said to do this specific thing, this was my opinion. I tested gcc and compiler explorer and didn't made a difference in this specific case at -O3)

amirroth replied:
I saw that too. I think it applies to header files only.

jmarrec commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -239,6 +228,10 @@ void GetPumpInput(EnergyPlusData &state)
+    constexpr std::array<std::string_view, static_cast<int>(PumpControlType::Num)> pumpCtrlTypeNamesUC{"CONTINUOUS", "INTERMITTENT"};
+    constexpr std::array<std::string_view, static_cast<int>(ControlTypeVFD::Num)> controlTypeVFDNamesUC{"MANUALCONTROL", "PRESSURESETPOINTCONTROL"};
+    constexpr std::array<std::string_view, static_cast<int>(PowerSizingMethod::Num)> powerSizingMethodNamesUC{"POWERPERFLOW",
+                                                                                                              "POWERPERFLOWPERPRESSURE"};
```
```suggestion
    static constexpr std::array<std::string_view, static_cast<int>(PumpControlType::Num)> pumpCtrlTypeNamesUC{"CONTINUOUS", "INTERMITTENT"};
    static constexpr std::array<std::string_view, static_cast<int>(ControlTypeVFD::Num)> controlTypeVFDNamesUC{"MANUALCONTROL", "PRESSURESETPOINTCONTROL"};
    static constexpr std::array<std::string_view, static_cast<int>(PowerSizingMethod::Num)> powerSizingMethodNamesUC{"POWERPERFLOW",
                                                                                                              "POWERPERFLOWPERPRESSURE"};
```

jmarrec commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -105,19 +105,8 @@ using DataHVACGlobals::ForceOff;
-    {PumpType::Bank_ConSpeed, cPumpBank_ConSpeed},
-};
+constexpr std::array<std::string_view, static_cast<int>(PumpType::Num)> pumpTypeIDFNames = {
+    "Pump:VariableSpeed", "Pump:ConstantSpeed", "Pump:VariableSpeed:Condensate", "HeaderedPumps:VariableSpeed", "HeaderedPumps:ConstantSpeed"};
```
```suggestion
static constexpr std::array<std::string_view, static_cast<int>(PumpType::Num)> pumpTypeIDFNames = {
    "Pump:VariableSpeed", "Pump:ConstantSpeed", "Pump:VariableSpeed:Condensate", "HeaderedPumps:VariableSpeed", "HeaderedPumps:ConstantSpeed"};
```

jmarrec commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -251,25 +244,23 @@ void GetPumpInput(EnergyPlusData &state)
     int NumVarSpeedPumps;
     int NumConstSpeedPumps;
     int NumCondensatePumps;
```
Not saying we need to do it, but IMHO declaration of variables at top of routine is a code smell.

*Uninitialized* declarations is even worse.

RKStrand replied:
Ok, they will be initialized.  At least it won't be "even worse" now.

jmarrec commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -281,950 +272,898 @@ void GetPumpInput(EnergyPlusData &state)
+    for (PumpNum = 1; PumpNum <= NumVarSpeedPumps; ++PumpNum) {
+        auto &thisPump = state.dataPumps->PumpEquip(PumpNum);
         state.dataInputProcessing->inputProcessor->getObjectItem(state,
                                                                  cCurrentModuleObject,
-                                                                 NumVarPump,
-                                                                 state.dataIPShortCut->cAlphaArgs,
+                                                                 PumpNum,
+                                                                 thisInput->cAlphaArgs,
                                                                  NumAlphas,
-                                                                 state.dataIPShortCut->rNumericArgs,
+                                                                 thisInput->rNumericArgs,
```
(Are we eventually moving to the epJSON stuff? eg stuff like here: https://github.com/NREL/EnergyPlus/blob/7401e1fee9615a0fc571718d1845c732182b8529/src/EnergyPlus/SystemAvailabilityManager.cc#L1302-L1343)

RKStrand replied:
I don't know.  Perhaps @amirroth can comment.  A lot of work has been done here already.  I don't know the plans long term for epJSON in EnergyPlus.

amirroth replied:
@jmarrec We are moving towards epJSON based inputs, but we are also potentially moving to a system in which some of that code will be automatically generated from the epJSON schema (a la OpenStudio). It doesn't make sense to continue doing this transition manually until we see what we can do automatically and what that actually looks like. And as @RKStrand mentioned, this was already a lot of surgery for a first time doing a pass like this.

jmarrec replied:
Totally agree this is completely out of scope. I put it inside parentheses but should probably have added a "Side note" disclaimer too :)

jmarrec commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -281,950 +272,898 @@ void GetPumpInput(EnergyPlusData &state)
-    for (NumVarPump = 1; NumVarPump <= NumVarSpeedPumps; ++NumVarPump) {
-        PumpNum = NumVarPump;
+    for (PumpNum = 1; PumpNum <= NumVarSpeedPumps; ++PumpNum) {
+        auto &thisPump = state.dataPumps->PumpEquip(PumpNum);
```
:+1: clears out so much text below...

jmarrec commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -281,950 +272,898 @@ void GetPumpInput(EnergyPlusData &state)
                 ShowSevereError(state,
-                                cCurrentModuleObject + "=\"" + state.dataIPShortCut->cAlphaArgs(1) + "\" invalid " +
-                                    state.dataIPShortCut->cAlphaFieldNames(7) + "=\"" + state.dataIPShortCut->cAlphaArgs(7) + "\" not found.");
+                                format("{}={} invalid {}={} not found.",
```
Is the removal of the `"` inside the format string wanted?

RKStrand replied:
Again, a matter of preference.  I can put them back in if that makes more sense to people.

jmarrec commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -281,950 +272,898 @@ void GetPumpInput(EnergyPlusData &state)
-                state.dataPumps->PumpEquip(PumpNum).powerSizingMethod = PowerSizingMethod::SizePowerPerFlowPerPressure;
-            } else {
+        if (!thisInput->lAlphaFieldBlanks(8)) {
+            thisPump.powerSizingMethod = static_cast<PowerSizingMethod>(
+                getEnumerationValue(powerSizingMethodNamesUC, UtilityRoutines::MakeUPPERCase(state.dataIPShortCut->cAlphaArgs(8))));
```
I've seen this idiom a lot until now... Should we make a higher level helper? it'll save one line, but may be clearer

```
template <typename T>
constexpr T getEnumFromString(const gsl::span<const std::string_view> sList,
                              const std::string_view s) {
  return static_cast<T>(getEnumerationValue(sList, s));
}
```
Usage : `thisPump.powerSizingMethod = getEnumFromString<PowerSizingMethod>(powerSizingMethodNamesUC, state.dataIPShortCut->cAlphaArgs(8));`

(not sure whether the MakeUPPERCase is needed nowadays or not, but that can be added)

RKStrand replied:
Good idea--I'd ask that we pin that and come back later.  There are lots of places in the code where this is going to be an issue.  Seems like it would be best if someone went through the code and dealt with these all at once.

amirroth replied:
Sure, why not.

Myoldmopar replied:
Yeah, using the template function to static cast it for us is a great idea.

jmarrec commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -1902,56 +1800,55 @@ void CalcPumps(EnergyPlusData &state, int const PumpNum, Real64 const FlowReques
-        if (DataPlant::CompData::getPlantComponent(state, state.dataPumps->PumpEquip(PumpNum).plantLoc).FlowCtrl ==
-            DataBranchAirLoopPlant::ControlType::SeriesActive) {
-            state.dataPumps->PumpMassFlowRate = 0.0;
+    if (thisPump.pumpType == PumpType::VarSpeed || thisPump.pumpType == PumpType::Bank_VarSpeed || thisPump.pumpType == PumpType::Cond) {
+        if (DataPlant::CompData::getPlantComponent(state, thisPump.plantLoc).FlowCtrl == DataBranchAirLoopPlant::ControlType::SeriesActive) {
+            daPumps->PumpMassFlowRate = 0.0;
```
I'm glad to see that BITF stuff gone. 👍

jmarrec commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -1233,169 +1172,155 @@ void GetPumpInput(EnergyPlusData &state)
         } break;
         default: {
-            ShowFatalError(state, format("Invalid Pump Type = {}", state.dataPumps->PumpEquip(PumpNum).pumpType));
+            assert(false);
```
missing `#include <cassert>` at beginning of file.

RKStrand replied:
I'll add it.

jmarrec commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -2366,59 +2208,49 @@ void ReportPumps(EnergyPlusData &state, int const PumpNum)
+    auto &thisOutNode = state.dataLoopNodes->Node(OutletNode);
+    auto &daPumps = state.dataPumps;
+
+    if (daPumps->PumpMassFlowRate <= DataBranchAirLoopPlant::MassFlowTolerance) {
+        new (&(state.dataPumps->PumpEquipReport(PumpNum))) ReportVars();
+        thisPumpRep.OutletTemp = thisOutNode.Temp;
```
I'm confused by this one (as in I'm not sure what that construct does). Could you explain the `new` line please? I'm getting lost.

RKStrand replied:
I think this basically resets the variable/structure back to its default value.  In this case, it zeroes everything out.  The code it replace did that with one exception (OutletTemp)., hence the next line.

amirroth replied:
This "new with placement". It's been a C++ feature since the beginning. By giving an address argument to `new`, you suppress the memory allocation and call the constructor (in this case `ReportVars()`) on the address provided. 

It's similar to `state.dataPumps->PumpEquipReport(PumpNum) = ReportVars();` but it doesn't construct and copy, it just constructs into the destination. Yes, I know about RVO. But why rely on RVO when you have this?

jmarrec replied:
Today I learned! Thanks

jmarrec reviewed (commented):
Left a couple of small comments. This looks much better!

RKStrand committed:
Address Addition Review Comments

Modification of several elements of the pump module based on review comments from @jmarrec.

RKStrand commented:
@jmarrec I have hopefully corrected all of the issues here other than the one I asked that we deal with later throughout the entire code.  These should all be in the commit that I made a little while ago.  If I missed anything, please let me know and I will correct that.

Myoldmopar commented on doc/input-output-reference/src/overview/group-pumps.tex:
```diff
@@ -64,7 +64,7 @@ \subsubsection{Inputs}\label{inputs-037}
 \paragraph{Field: Design Minimum Flow Rate}\label{field-design-minimum-flow-rate}
 
-This field contains the minimum volume flow rate while operating in variable flow capacity rate in cubic meters per second. This field is autosizable. When set to autosize the value in the input field Design Minimum Flow Rate Sizing Factor is used to scale the minimum flow rate as a fraction of the Design Maximum Flow Rate.
+This field contains the minimum volume flow rate while operating in variable flow capacity rate in cubic meters per second. This field is autosizable. When set to autosize the value in the input field Design Minimum Flow Rate Sizing Factor is used to scale the minimum flow rate as a fraction of the Design Maximum Flow Rate.  Note that this Design Minimum Flow Rate should be less than the Design M ...
```
OK

Myoldmopar commented on src/EnergyPlus/Pumps.hh:
```diff
@@ -48,9 +48,6 @@
 #define Pumps_hh_INCLUDED
 
-// ObjexxFCL Headers
-#include <ObjexxFCL/Array1D.hh>
```
Nice.

Myoldmopar commented on src/EnergyPlus/Pumps.hh:
```diff
@@ -93,12 +90,6 @@ namespace Pumps {
-    extern std::string const cPump_ConSpeed;
-    extern std::string const cPump_Cond;
-    extern std::string const cPumpBank_VarSpeed;
-    extern std::string const cPumpBank_ConSpeed;
```
Yes.

Myoldmopar commented on src/EnergyPlus/Pumps.hh:
```diff
@@ -122,131 +113,93 @@ namespace Pumps {
+        bool HeatLossesToZone = false;        // if true then pump losses added to surrounding zone
+        int ZoneNum = 0;                      // index for zone surrounding pump
+        Real64 SkinLossRadFraction = 0.0;     // radiative split for skin losses to zone
+        bool LoopSolverOverwriteFlag = false; // loop solver overwrite for determining pump minimum flow rate
```
Much cleaner

Myoldmopar commented on src/EnergyPlus/Pumps.hh:
```diff
@@ -122,131 +113,93 @@ namespace Pumps {
+        Real64 ZoneTotalGainRate = 0.0;     // total pump skin losses to zone (W)
+        Real64 ZoneTotalGainEnergy = 0.0;   // total pump skin losses to zone energy (J)
+        Real64 ZoneConvGainRate = 0.0;      // pump skin losses convecting to zone air (W)
+        Real64 ZoneRadGainRate = 0.0;       // pump skin losses radiating to inside of zone (W)
```
One possible addition is to just eliminate this extra report vars struct entirely.  The variables can live on the base pump data structure, as they are both allocated to the same size every single time anyway.  No big deal, just noting for the future.

Myoldmopar commented on src/EnergyPlus/Pumps.hh:
```diff
@@ -292,8 +245,8 @@ struct PumpsData : BaseGlobalStruct
     Real64 PumpHeattoFluid = 0.0;  // Pump Power dissipated in fluid stream
     Real64 Power = 0.0;            // Pump Electric power
     Real64 ShaftPower = 0.0;       // Power passing through pump shaft
```
This data structure needs to be cleaned out.  There is no reason inside EnergyPlus that we need a global-ish variable to hold `Power`.  Each pump can hold its own value of power consumption.  Whenever this change is made, it needs to be made cautiously because it can easily lead to diffs.

Myoldmopar commented on tst/EnergyPlus/unit/Pumps.unit.cc:
```diff
@@ -651,8 +651,8 @@ TEST_F(EnergyPlusFixture, VariableSpeedPump_MinFlowGreaterThanMax)
         "   ** Warning ** GetPumpInput: Pump:VariableSpeed=\"SUPPLY INLET PUMP\", Invalid 'Design Minimum Flow Rate'",
-        "   **   ~~~   ** Entered Value=[2.00000E-003] is above the Design Maximum Flow Rate=[1.00000E-003].",
-        "   **   ~~~   ** Reseting value of 'Design Minimum Flow Rate' to the value of 'Design Maximum Flow Rate'.",
+        "   **   ~~~   ** Entered Value=[2.00000E-003] is above or too close (equal) to the Design Maximum Flow Rate=[1.00000E-003].",
```
:+1:

Myoldmopar commented on tst/EnergyPlus/unit/Pumps.unit.cc:
```diff
@@ -661,4 +661,59 @@ TEST_F(EnergyPlusFixture, VariableSpeedPump_MinFlowGreaterThanMax)
+    EXPECT_NEAR(state->dataPumps->PumpEquip(1).MinVolFlowRate, expectedAnswer, allowableTolerance);
+    expectedAnswer = 0.001;
+    EXPECT_NEAR(state->dataPumps->PumpEquip(1).NomVolFlowRate, expectedAnswer, allowableTolerance);
+}
```
Yes, great unit test!

Myoldmopar commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -105,19 +106,8 @@ using DataHVACGlobals::ForceOff;
-    {PumpType::Bank_VarSpeed, cPumpBank_VarSpeed},
-    {PumpType::Bank_ConSpeed, cPumpBank_ConSpeed},
-};
+static constexpr std::array<std::string_view, static_cast<int>(PumpType::Num)> pumpTypeIDFNames = {
```
Yep.

Myoldmopar commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -281,950 +274,907 @@ void GetPumpInput(EnergyPlusData &state)
+        TestCompSet(state, cCurrentModuleObject, thisPump.Name, thisInput->cAlphaArgs(2), thisInput->cAlphaArgs(3), "Water Nodes");
+
+        thisPump.PumpControl = static_cast<PumpControlType>(
+            getEnumerationValue(pumpCtrlTypeNamesUC, UtilityRoutines::MakeUPPERCase(state.dataIPShortCut->cAlphaArgs(4))));
```
Nice.

Myoldmopar commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -281,950 +274,907 @@ void GetPumpInput(EnergyPlusData &state)
+            }
+        }
+
+        thisPump.NomVolFlowRate = thisInput->rNumericArgs(1);
```
Interesting, I hadn't thought of aliasing the DataIPShortcut.  I would suggest that in the future, instead of doing that, just remove the usage of DataIPShortcut.  The purpose of that module was to hold a set of global arrays that each developer could use without having to manage and resize arrays themselves.  In C++, we can just pass vectors in and out of the input processor routines without having to presize them.  Then you can just use local variables named rNumericArgs, etc.  And there won't be any need to alias anything.

Myoldmopar commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -281,950 +274,907 @@ void GetPumpInput(EnergyPlusData &state)
+        thisPump.MinVolFlowRate = thisInput->rNumericArgs(10);
+        if (thisPump.MinVolFlowRate == AutoSize) {
+            thisPump.minVolFlowRateWasAutosized = true;
+        } else if (!thisPump.NomVolFlowRateWasAutoSized && (thisPump.MinVolFlowRate > (minToMaxRatioMax * thisPump.NomVolFlowRate))) {
```
New condition here, got it.

Myoldmopar reviewed (approved):
Good to go.

Myoldmopar commented:
This is good to go in, CI is good and everything is happy locally as well.
